eclair-core/src/main/scala/fr/acinq/eclair/channel/publish/ReplaceableTxPublisher.scala

```diff
@@ -73,485 +75,270 @@ object ReplaceableTxPublisher {
+  def fund(replyTo: ActorRef[TxPublisher.PublishTxResult], cmd: TxPublisher.PublishReplaceableTx, txWithWitnessData: ReplaceableTxWithWitnessData): Behavior[Command] = {
+    val targetFeerate = getFeerate(nodeParams.onChainFeeConf.feeEstimator, cmd.txInfo.confirmBefore, BlockHeight(nodeParams.currentBlockHeight))
+    val txFunder = context.spawn(ReplaceableTxFunder(nodeParams, bitcoinClient, loggingInfo), "tx-funder")
+    txFunder ! ReplaceableTxFunder.FundTransaction(context.messageAdapter[ReplaceableTxFunder.FundingResult](WrappedFundingResult), cmd, Right(txWithWitnessData), targetFeerate)
     Behaviors.receiveMessagePartial {
-      case TimeLocksOk =>
-        val htlcFeerate = cmd.commitments.localCommit.spec.htlcTxFeerate(cmd.commitments.commitmentFormat)
-        if (targetFeerate <= htlcFeerate) {
-          val channelKeyPath = keyManager.keyPath(cmd.commitments.localParams, cmd.commitments.channelConfig)
-          val localPerCommitmentPoint = keyManager.commitmentPoint(channelKeyPath, cmd.commitments.localCommit.index)
-          val localHtlcBasepoint = keyManager.htlcPoint(channelKeyPath)
-          val localSig = keyManager.sign(htlcTxWithWitnessData.txInfo, localHtlcBasepoint, localPerCommitmentPoint, TxOwner.Local, cmd.commitments.commitmentFormat)
-          val signedHtlcTx = htlcTxWithWitnessData.addSigs(localSig, cmd.commitments.commitmentFormat)
-          log.info("publishing {} without adding inputs: txid={}", cmd.desc, signedHtlcTx.tx.txid)
-          publish(replyTo, cmd, signedHtlcTx.tx, signedHtlcTx.fee)
-        } else {
-          fund(replyTo, cmd, targetFeerate)
+      case WrappedFundingResult(result) =>
+        result match {
+          case success: ReplaceableTxFunder.TransactionReady => publish(replyTo, cmd, success.fundedTx)
+          case ReplaceableTxFunder.FundingFailed(reason) => sendResult(replyTo, TxPublisher.TxRejected(loggingInfo.id, cmd, reason))
         }
       case Stop =>
-        timeLocksChecker ! TxTimeLocksMonitor.Stop
-        Behaviors.stopped
+        // We can't stop right now, the child actor is currently funding the transaction and will send its result soon.
+        // We just wait for the funding process to finish before stopping (in the next state).
+        timers.startSingleTimer(Stop, 1 second)
+        Behaviors.same
     }
   }
 
-  def checkTimeLocks(replyTo: ActorRef[TxPublisher.PublishTxResult], cmd: TxPublisher.PublishReplaceableTx, claimHtlcTx: ClaimHtlcTx, targetFeerate: FeeratePerKw): Behavior[Command] = {
-    val timeLocksChecker = context.spawn(TxTimeLocksMonitor(nodeParams, watcher, loggingInfo), "time-locks-monitor")
-    timeLocksChecker ! CheckTx(context.messageAdapter[TxTimeLocksMonitor.TimeLocksOk](_ => TimeLocksOk), cmd.txInfo.tx, cmd.desc)
+  def publish(replyTo: ActorRef[TxPublisher.PublishTxResult], cmd: TxPublisher.PublishReplaceableTx, tx: FundedTx): Behavior[Command] = {
```

Comment:
You can't do it easily, because you want to register to the `CurrentBlockCount` event only once, whereas we go back to the `wait` state regularly (whenever we fee-bump). I could add a flag `registerToBlockCount` or something like it and wrap that part in an `if`, but that feels less clean...

Reply:
We could register `CurrentBlockCount` once in `fund()` before calling `publish()` the first time? Note that with https://github.com/ACINQ/eclair/pull/2113#discussion_r784119211 we don't need to register to `CurrentBlockCount`.

Reply:
Fixed in https://github.com/ACINQ/eclair/pull/2113/commits/8d13b911e5c6743bdc9c51909e9901bc093adb04